pkg/abcupdater/metrics.go

```diff
@@ -0,0 +1,226 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package abcupdater
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/sethvargo/go-envconfig"
+
+	"github.com/abcxyz/abc-updater/pkg/abcupdater/localstore"
+	"github.com/abcxyz/pkg/logging"
+)
+
+const (
+	installIDFileName = "id.json"
+)
+
+var regExUUID = regexp.MustCompile("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
+
+type metricsConfig struct {
+	ServerURL string `env:"ABC_UPDATER_METRICS_URL,default=https://abc-updater-metrics.tycho.joonix.net"`
+}
+
+// MetricsInfo are the parameters for sending metrics.
+type MetricsInfo struct {
+	// The ID of the application to check.
+	AppID string
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string
+
+	// An optional Lookuper to load envconfig structs. Will default to os environment variables.
+	Lookuper envconfig.Lookuper
+
+	Metrics map[string]int
+
+	// Optional override for install id file location. Mostly intended for testing.
+	// If empty uses default location.
+	InstallIDFileOverride string
+}
+
+// InstallIDData defines the json file that defines installation id.
+type InstallIDData struct {
+	// Time ID was created, in UTC epoch seconds.
+	IDCreatedTimestamp int64 `json:"idCreatedTimestamp"`
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+type SendMetricRequest struct {
+	// The ID of the application to check.
+	AppID string `json:"appId"`
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string `json:"version"`
+
+	Metrics map[string]int `json:"metrics"`
+
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+// Stricter than uuid.Parse() which isn't meant for validating strings according
+// to documentation.
+func validInstallID(id string) bool {
+	return regExUUID.MatchString(id)
+}
+
+// SendMetricsSync sends information about application usage. Users can opt out
+// by setting an env variable defined in opt_out.go.
+// Accepts a context for cancellation.
+func SendMetricsSync(ctx context.Context, info *MetricsInfo) error {
+	lookuper := info.Lookuper
+	if lookuper == nil {
+		lookuper = envconfig.OsLookuper()
+	}
```

Comment:
Similar to my comment above, this is a bit strange. I don't like coupling the input `MetricsInfo` to the way the environment is resolved. It would be better to create a dedicated structure for sending metrics (which I think will be useful in the long run anyway) and have the lookuper initialized there.

```go
type MetricsClient struct {
  lookuper envconfig.Lookuper
}

func NewMetricsClient(opts ...Option) *MetricsClient {
  // ...
}

func (c *MetricsClient) SendSync(ctx context.Context, info *MetricsInfo) error {
  c.lookeruper.Lookup()
}
```

That would also let you load the opt-out settings once globally instead of per-metric.